### ISSUE TITLE
Python wrapper error: parallel_for failed: out of memory

### ISSUE THREAD

**Issue by ktamiola** [CannyLab/tsne-cuda#12]
**Error**
```
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  parallel_for failed: out of memory
Aborted (core dumped)
```

**Code to reproduce**
```Python
from tsnecuda import TSNE
import numpy as np

# Just a data limit
N = 1000000
iter = 1000

X = np.random.rand(N, 2)

X_embedded = TSNE(n_components=2, verbose=1, n_iter=iter).fit_transform(X)

# Print the frame
print(X_embedded)
```

**Configuration**
```
GPU:
Tesla V100-SXM2-16gb
NVIDIA Driver Version: 396.37
```

**Compiled from source with**
```
g++ (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2017 NVIDIA Corporation
Built on Fri_Sep__1_21:08:03_CDT_2017
Cuda compilation tools, release 9.0, V9.0.176
```

**Comment by DavidMChan** [comment]
Does the executable version fail with 1,000,000 points as well? That is, can you run:
```
./tsne -k 1000000
```
from the build directory?

**Comment by ktamiola** [comment]
I can successfully execute. 
```
./tsne -k 1000000
```
Thus the issue must be in the Python wrapper.

**Comment by ktamiola** [comment]
Excellent suggestion. Indeed, decreasing `num_neighbors` to `32` does the trick. 

However, there seem to be a serious architecture optimization issue + a memory allocation shenanigan. 

```
python2.7 tsne-gpu.py
[Step 0] Avg. Gradient Norm: 5.11523e-07
[Step 10] Avg. Gradient Norm: 0.000230085
[Step 20] Avg. Gradient Norm: 0.000718175
[Step 30] Avg. Gradient Norm: 0.00122661
[Step 40] Avg. Gradient Norm: 0.00168105
[Step 50] Avg. Gradient Norm: 0.00245505
[Step 60] Avg. Gradient Norm: 0.00464572
[Step 70] Avg. Gradient Norm: 0.0101716
[Step 80] Avg. Gradient Norm: 0.0197001
[Step 90] Avg. Gradient Norm: 0.0269809
[Step 100] Avg. Gradient Norm: 0.0281236
[Step 110] Avg. Gradient Norm: 0.026937
[Step 120] Avg. Gradient Norm: 0.025774
[Step 130] Avg. Gradient Norm: 0.0250006
[Step 140] Avg. Gradient Norm: 0.0245269
GPUassert: an illegal memory access was encountered /home/ubuntu/tsne-cuda/src/util/cuda_utils.cu 55
```  
I am trying to calculate the tSNE for a  50000 x 2  `numpy` array, using 32 neighbours.

**Comment by ktamiola** [comment]
When I boos the array size to 500,00 tSNE only completes the first step
```
Reached minimum gradient norm: 5.28734e-08
```
and exits without a noticeable error.

**Comment by DavidMChan** [comment]
The second issue is caused because the minimum gradient norm is set too high for your dataset (so the early termination kills the process after a single step). Decreasing the min gradient norm will allow the program to continue. 

 The former error is one which we are tracking, and is likely caused by two nodes being pushed close enough together that they reside in the same location, causing instability in the program. Both decreasing the learning rate, and increasing the perplexity can alleviate this issue.  We've been trying to find the bug for a while, and we have no clue where it is - but it's on our active todo list.

**Comment by ktamiola** [comment]
Understood. 🤦‍♂️ I should have read warning message with understanding (it is late here in EU, now).

**Comment by ktamiola** [comment]
Just to confirm, I have launched tSNE for 4,000,000 x 2 points using the Python wrapper. All seems stable.

**Comment by DavidMChan** [comment]
Mostly resolved with the exception of issues #17.
